photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/AbstractOptimizer.scala

```diff
@@ -44,7 +44,17 @@ abstract class AbstractOptimizer[Datum <: DataPoint, -Function <: DiffFunction[D
     None
   }
 
-  def isDone: Boolean = convergenceReason.nonEmpty
+  def isDone: Boolean = {
+    val isDone = convergenceReason.nonEmpty
+    if (isDone) {
+      // Set the convergenceReason when the optimization is done
+      statesTracker match {
+        case Some(y) => y.convergenceReason = convergenceReason
+        case None =>
+      }
+    }
+    isDone
+  }
```

Comment:
Separated set convergence reason from set current/previous states logic.

For now we can't really make too much code change for ```AbstractOptimizer``` because the unit tests are not there, so I opened #104 to remind us this issue.